documentloaders/cloudsql/documentloader.go

```diff
@@ -0,0 +1,363 @@
+	return &valueProcessor{
+		extractors: map[reflect.Type]func(any) (any, bool){
+			reflect.TypeOf(&sql.NullTime{}): func(v any) (any, bool) {
+				if nv, ok := v.(*sql.NullTime); ok && nv.Valid {
+					return nv.Time, true
+				}
+				return nil, false
+			},
+			reflect.TypeOf(&sql.NullString{}): func(v any) (any, bool) {
+				if nv, ok := v.(*sql.NullString); ok && nv.Valid {
+					return nv.String, true
+				}
+				return nil, false
+			},
+			reflect.TypeOf(&sql.NullBool{}): func(v any) (any, bool) {
+				if nv, ok := v.(*sql.NullBool); ok && nv.Valid {
+					return nv.Bool, true
+				}
+				return nil, false
+			},
+			reflect.TypeOf(&sql.NullInt64{}): func(v any) (any, bool) {
+				if nv, ok := v.(*sql.NullInt64); ok && nv.Valid {
+					return nv.Int64, true
+				}
+				return nil, false
+			},
+			reflect.TypeOf(&sql.NullFloat64{}): func(v any) (any, bool) {
+				if nv, ok := v.(*sql.NullFloat64); ok && nv.Valid {
+					return nv.Float64, true
+				}
+				return nil, false
+			},
+			reflect.TypeOf(&sql.RawBytes{}): func(v any) (any, bool) {
+				if rv, ok := v.(*sql.RawBytes); ok {
+					return *rv, true
+				}
+				return nil, false
+			},
+		},
+	}
+}
+
+func (vp *valueProcessor) process(valuesPrt []any, columnNames []string) map[string]any {
+	columnValues := make(map[string]any)
+	for i, name := range columnNames {
+		if handler, ok := vp.extractors[reflect.TypeOf(valuesPrt[i])]; ok {
+			if val, valid := handler(valuesPrt[i]); valid {
+				columnValues[name] = val
+				continue
+			}
+		}
+		columnValues[name] = valuesPrt[i]
+	}
+	return columnValues
+}
+
+type oidProcessor struct {
+	extractors map[pgtype.OID]any
+}
+
+func newOidProcessor() *oidProcessor {
+	return &oidProcessor{
+		extractors: map[pgtype.OID]any{
+			pgtype.TimeOID:        new(sql.NullTime),
+			pgtype.TimestampOID:   new(sql.NullTime),
+			pgtype.TimestamptzOID: new(sql.NullTime),
+			pgtype.DateOID:        new(sql.NullTime),
+			pgtype.VarcharOID:     new(sql.NullString),
+			pgtype.TextOID:        new(sql.NullString),
+			pgtype.JSONOID:        new(sql.NullString),
+			pgtype.BoolOID:        new(sql.NullBool),
+			pgtype.Float4OID:      new(sql.NullFloat64),
+			pgtype.Float8OID:      new(sql.NullFloat64),
+			pgtype.Int2OID:        new(sql.NullInt64),
+			pgtype.Int4OID:        new(sql.NullInt64),
+			pgtype.Int8OID:        new(sql.NullInt64),
+		},
+	}
+}
+
+func (vp *oidProcessor) process(oid pgtype.OID) (any, bool) {
+	v, ok := vp.extractors[oid]
+	return v, ok
+}
+
+// DocumentLoader is responsible for loading documents from a Postgres database.
+type DocumentLoader struct {
+	engine             cloudsqlutil.PostgresEngine
+	query              string
+	tableName          string
+	schemaName         string
+	contentColumns     []string
+	metadataColumns    []string
+	metadataJSONColumn string
+	format             string
+	formatter          func(map[string]any, []string) string
+
+	oidProcessor   *oidProcessor
+	valueProcessor *valueProcessor
```

Comment:
Would love some comments to help on the fields who's use is non-obvious. For example, I'm not sure I understand how format or formatter are applied here?